src/backend/distributed/planner/multi_planner.c

```diff
@@ -122,38 +124,152 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
+		commandType == CMD_DELETE || query->hasModifyingCTE)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * CreateDistributedPlan encapsulates the logic needed to transform a particular
+ * query into a distributed plan.
+ */
+static PlannedStmt *
+CreateDistributedPlan(PlannedStmt *localPlan, Query *originalQuery, Query *query,
+					  ParamListInfo boundParams,
+					  RelationRestrictionContext *restrictionContext)
+{
+	MultiPlan *distributedPlan = NULL;
+	PlannedStmt *resultPlan = NULL;
+	bool hasUnresolvedParams = false;
+
+	if (HasUnresolvedExternParamsWalker((Node *) query, boundParams))
+	{
+		hasUnresolvedParams = true;
+	}
+
+	if (IsModifyCommand(query))
+	{
+		/* modifications are always routed through the same planner/executor */
+		distributedPlan = CreateModifyPlan(originalQuery, query, restrictionContext);
+		Assert(distributedPlan);
+	}
+	else
 	{
-		/* Create and optimize logical plan */
-		MultiTreeRoot *logicalPlan = MultiLogicalPlanCreate(query);
-		MultiLogicalPlanOptimize(logicalPlan);
+		/*
+		 * For select queries we, if router executor is enabled, first try to
+		 * plan the query as a router query. If not supported, otherwise try
+		 * the full blown plan/optimize/physical planing process needed to
+		 * produce distributed query plans.
+		 */
+		if (EnableRouterExecution)
+		{
+			distributedPlan = CreateRouterPlan(originalQuery, query, restrictionContext);
+
+			/* for debugging it's useful to display why query was not router plannable */
+			if (distributedPlan && distributedPlan->planningError)
+			{
+				RaiseDeferredError(distributedPlan->planningError, DEBUG1);
+			}
+		}
 
 		/*
-		 * This check is here to make it likely that all node types used in
-		 * Citus are dumpable. Explain can dump logical and physical plans
-		 * using the extended outfuncs infrastructure, but it's infeasible to
-		 * test most plans. MultiQueryContainerNode always serializes the
-		 * physical plan, so there's no need to check that separately.
+		 * Router didn't yield a plan, try the full distributed planner. As
+		 * real-time/task-tracker don't support prepared statement parameters,
+		 * skip planning in that case (we'll later trigger an error in that
+		 * case if necessary).
 		 */
-		CheckNodeIsDumpable((Node *) logicalPlan);
+		if ((!distributedPlan || distributedPlan->planningError) && !hasUnresolvedParams)
+		{
+			/* Create and optimize logical plan */
+			MultiTreeRoot *logicalPlan = MultiLogicalPlanCreate(query);
+			MultiLogicalPlanOptimize(logicalPlan);
+
+			/*
+			 * This check is here to make it likely that all node types used in
+			 * Citus are dumpable. Explain can dump logical and physical plans
+			 * using the extended outfuncs infrastructure, but it's infeasible to
+			 * test most plans. MultiQueryContainerNode always serializes the
+			 * physical plan, so there's no need to check that separately.
+			 */
+			CheckNodeIsDumpable((Node *) logicalPlan);
```

Comment:
Yes, I think so - I'd at some point decided to remove the unused output functions for the logical stage, but several people including @metdos and @ozgune said they'd like to keep them.  Having some way to exercise nearly all of them sounds like a good idea.